torch/distributed/_fsdp/wrap.py

```diff
@@ -161,16 +161,89 @@ def wrap(module: nn.Module, **wrap_overrides: Any) -> nn.Module:
+    # kwargs passed into config auto wrap, then module.wrapper_config kwargs,
+    # then kwargs passed into wrap(), from least to highest precedence. Thus,
+    # ConfigAutoWrap.wrap constructs according to this priority and skips
+    # below logic by passing check_module_overrides=False.
+    if check_module_overrides and hasattr(module, "wrapper_config"):
+        module_overrides = module.wrapper_config
+        assert isinstance(module_overrides, dict)
+    wrap_overrides = {**kwargs, **module_overrides}
+    assert wrapper_cls is not None
+    return wrapper_cls(module, **wrap_overrides)
+
+
+def recursive_wrap(
+    module: nn.Module,
+    auto_wrap_policy: Callable,
+    wrapper_cls: Callable,
+    only_wrap_children: bool = False,
+    **kwargs: Any
+) -> Tuple[nn.Module, int]:
+    """
+    Automatically wrap child modules of *module* that meet the given
+    criteria with :func:`auto_wrap`. Does not rely on ConfigAutoWrap.
+    Args:
+        module (nn.Module):
+            module to recursively wrap
+        auto_wrap_policy (Callable):
+            A callable specifying a policy to recursively wrap layers with FSDP.
+    Returns:
+        (nn.Module, int):
+            Wrapped module and the number parameters wrapped recursively.
+    """
+    assert auto_wrap_policy is not None, "Must specify auto_wrap_policy."
+    assert wrapper_cls is not None, "Must specify wrapper_cls"
+    # Make sure no child is already wrapped.
+    for _, child in module.named_modules():
+        assert not isinstance(child, cast(type, wrapper_cls))
+
+    # We count all params, assuming none of them is already wrapped.
+    num_params = sum([p.numel() for p in module.parameters()])
+
+    assert auto_wrap_policy is not None
+    if auto_wrap_policy(module=module, recurse=True, unwrapped_params=num_params):
+        total_wrapped_params = 0
+        # Iterate through the children, recursively wrap if necessary
+        for name, child in module.named_children():
+            wrapped_child, num_wrapped_params = recursive_wrap(
+                module=child, auto_wrap_policy=auto_wrap_policy, wrapper_cls=wrapper_cls, **kwargs
+            )
+            setattr(module, name, wrapped_child)
+            # Keep track of how many parameters have been wrapped
+            total_wrapped_params += num_wrapped_params
+        # decide if we need to wrap the current module,
+        # since the left over parameters exceed the number of params to wrap
+        remainder = num_params - total_wrapped_params
+        if not only_wrap_children and auto_wrap_policy(module=module, recurse=False, unwrapped_params=remainder):
+            # Leaf node or final wrapping of the remainder both happen here.
+            return _wrap(module, wrapper_cls, **kwargs), num_params
+        else:
+            return module, total_wrapped_params
+    return module, 0
+
+
 def auto_wrap(module: nn.Module, auto_wrap_policy: Optional[Callable] = None, **kwargs: Any) -> nn.Module:
```

Comment:
since no one is using this, we can just kill it?

Reply:
I think right now our internal benchmarks are using it, I was going to remove it from those and then remove these APIs

Reply:
benchmarks are not actually using it so i removed it in the above diffs